File: src/integrator/ExtPlumed.py
#  Copyright (C) 2018
#      Max Planck Institute for Polymer Research
#
#  This file is part of ESPResSo++.
#
#  ESPResSo++ is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  ESPResSo++ is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.


r"""
*************************************************
espressopp.integrator.ExtPlumed
*************************************************

This extension serves as interface between ESPResSo++ and `PLUMED <http://www.plumed.org/home>`_. PLUMED can
be used to run realtime analysis and bias the simulation along chosen collective variables. Details of using
PLUMED and its input file can be found on its website.

To call PLUMED, set environment variable :envvar:`PLUMED_KERNEL` to the full path of libplumedKernel.so, and add
the path of the directory of libplumed.so to environment variable :envvar:`LD_LIBRARY_PATH`, when PLUMED is linked through runtime linking.
If PLUMED is linked by dynamic linking, i.e., it is linked as a shared library to ESPResSo++, :envvar:`PLUMED_KERNEL` needs not to be set.
If PLUMED is linked statically, neither environment variable needs to be provided.
Error messages regarding to "undefined symbol" can be ignored, as long as the simulation runs.

By default, PLUMED is linked by runtime linking. The user can change the way of PLUMED linking to ESPResSo++ by changing CMake variable :makevar:`PLUMED_LINK_TYPE`
to either "shared" or "static" from the default "runtime". If your distribution of PLUMED is not installed in a default system directory, you can define the root
path of PLUMED installation as CMake variable :makevar:`PLUMED_HOME`. If you keep the default "runtime" linking, your installation of PLUMED is not searched. Hence, you
do not need to worry about whether your PLUMED is not installed in a default system directory or setting :makevar:`PLUMED_HOME`.

usage:

.. code:: python

    plumed = espressopp.integrator.ExtPlumed(system, "plumed.dat", "log.plumed", 0.005)
    plumed.setNaturalUnits()
    plumed.Init()
    integrator.addExtension(plumed)

or:

.. code:: python

    plumed = espressopp.integrator.ExtPlumed(system, "plumed.dat", "log.plumed", 0.005)
    plumed.setNaturalUnits()
    plumed.setRestart(1)
    plumed.Init()
    integrator.addExtension(plumed)

.. function:: espressopp.integrator.ExtPlumed(system, cmd, log, dt)

		:param system: The Espresso++ system object.
                :type system: espressopp.System
                :param cmd: input file for PLUMED
                :type cmd: ``str``
                :param log: log file for PLUMED
                :type log:  ``str``
                :param dt:  time step
                :type dt: ``float`` (default: 0.005)

"""
from espressopp.esutil import cxxinit
from espressopp import pmi
from espressopp.integrator.Extension import *
from _espressopp import integrator_ExtPlumed
import mpi4py.MPI as MPI

class ExtPlumedLocal(ExtensionLocal, integrator_ExtPlumed):

    def __init__(self, system, cmd, log, dt=0.005):
        if pmi._PMIComm and pmi._PMIComm.isActive():
            if pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
                cxxinit(self, integrator_ExtPlumed, system, pmi._PMIComm.getMPIsubcomm(), cmd, log, dt)
            else:
                pass
        else:
            cxxinit(self, integrator_ExtPlumed, system, pmi._MPIcomm, cmd, log, dt)

if pmi.isController :
    class ExtPlumed(Extension):
        __metaclass__ = pmi.Proxy
        pmiproxydefs = dict(
            cls = 'espressopp.integrator.ExtPlumedLocal',
            pmicall = ['getBias', 'setNaturalUnits', 'setTimeUnit', 'setEnergyUnit', 'setLengthUnit', 'setKbT', 'setRealPrecison', 'setMDChargeUnit', 'setMDMassUnit', 'setRestart', 'Init']
            )

File: src/integrator/ExtPlumed.cpp
/*
  Copyright (C) 2018
  Max Planck Institute for Polymer Research

  This file is part of ESPResSo++.

  ESPResSo++ is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  ESPResSo++ is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <string>
#include "python.hpp"
#include "System.hpp"
#include "storage/Storage.hpp"
#include "interaction/Interaction.hpp"
#include "iterator/CellListIterator.hpp"
#include "bc/BC.hpp"
#include "ExtPlumed.hpp"
#include "mpi.hpp"

namespace espressopp {

  using namespace iterator;
  using std::string;

  namespace integrator {

    ExtPlumed::ExtPlumed(shared_ptr<System> _system, python::object _pyobj, string _dat, string _log, real _dt, bool _restart):
      Extension(_system),
      dt(_dt),
      step(0),
      nreal(0),
      gatindex(NULL),
      masses(NULL),
      f(NULL),
      pos(NULL),
      charges(NULL),
      particlesChanged(false)
    {
      p=new PLMD::Plumed;
      int real_precision = sizeof(real);
      p->cmd("setRealPrecision",&real_precision);
      p->cmd("setMDEngine","ESPResSo++");
      MPI_Comm comm = MPI_Comm(*_system->comm);
      p->cmd("setMPIComm", &comm);
      bool dat_is_file = (_dat.find_first_of("\n") > _dat.size()); // test if input is a multline string.
      if (dat_is_file) p->cmd("setPlumedDat", _dat.c_str());
      p->cmd("setLogFile", _log.c_str());
      p->cmd("setTimestep",&dt);
      longint tmp = _system->storage->getNRealParticles();
      boost::mpi::all_reduce(*_system->comm, tmp, natoms, std::plus<longint>());
      p->cmd("setNatoms",&natoms);
      if (_restart) {
        int res = 1;
        p->cmd("setRestart", &res);
      }
      p->cmd("init");
      if (!dat_is_file) {
        std::istringstream iss(_dat);
        std::string token;
        iss >> std::ws; // remove leading white spaces in a line
        while(std::getline(iss, token)) {
          iss >> std::ws; // remove leading white spaces in a line
          if (token[0] == '#') continue;
          p->cmd("readInputLine", token.c_str());
          token.clear();
        }
      }
      _onParticlesChanged = _system->storage->onParticlesChanged.connect(boost::bind(&ExtPlumed::onParticlesChanged, this));
    }

    ExtPlumed::~ExtPlumed() {
      _onParticlesChanged.disconnect();
      delete [] f;
      delete [] pos;
      delete [] charges;
      delete [] gatindex;
      delete [] masses;
      delete p;
    }

    real ExtPlumed::getBias() {
      return bias;
    }

    void ExtPlumed::onParticlesChanged() {
      particlesChanged = true;
    }

    void ExtPlumed::disconnect() {
      _runInit.disconnect();
      _aftCalcF.disconnect();
      _aftIntP.disconnect();
    }

    void ExtPlumed::connect() {
      _runInit = integrator->runInit.connect( boost::bind(&ExtPlumed::setStep, this));
      _aftCalcF = integrator->aftCalcF.connect( boost::bind(&ExtPlumed::updateForces, this));
      _aftIntP = integrator->aftIntP.connect( boost::bind(&ExtPlumed::updateStep, this));
    }

    void ExtPlumed::setStep() {
      step = integrator->getStep();
    }

    void ExtPlumed::updateForces() {
      System& system = getSystemRef();
      CellList realCells = system.storage->getRealCells();

      if (nreal!=system.storage->getNRealParticles()) {
        nreal = system.storage->getNRealParticles();
        if(charges) delete [] charges;
        if(masses) delete [] masses;
        if(gatindex) delete [] gatindex;
        if(pos) delete [] pos;
        if(f) delete [] f;
        gatindex = new int [nreal];
        masses = new real [nreal];
        charges = new real [nreal];
        pos = new real[nreal*3];
        f = new real[nreal*3];

        for(auto tp=std::make_pair(0, CellListIterator(realCells));
            tp.first<nreal && !tp.second.isDone();
            ++tp.first, ++tp.second)
          {
            gatindex[tp.first] = static_cast<int>(tp.second->id())-1;
          }
        particlesChanged = false;

      } else if (particlesChanged) {
        for(auto tp=std::make_pair(0, CellListIterator(realCells));
            tp.first<nreal && !tp.second.isDone();
            ++tp.first, ++tp.second)
          {
            gatindex[tp.first] = static_cast<int>(tp.second->id())-1;
          }
        particlesChanged = false;
      }

      for(auto tp=std::make_pair(0, CellListIterator(realCells));
          tp.first<nreal && !tp.second.isDone();
          ++tp.first, ++tp.second)
        {
          masses[tp.first] = tp.second->mass();
          charges[tp.first] = tp.second->q();
          std::copy(tp.second->force().begin(), tp.second->force().end(), &f[tp.first*3]);
          std::copy(tp.second->position().begin(), tp.second->position().end(), &pos[tp.first*3]);
        }

      p->cmd("setStep",&step);
      p->cmd("setAtomsNlocal",&nreal);
      p->cmd("setAtomsGatindex",&gatindex[0]);

      real box[3][3];
      for(int i=0;i<3;i++) for(int j=0;j<3;j++) box[i][j]=0.0;
      Real3D L = system.bc->getBoxL();
      box[0][0]=L[0];
      box[1][1]=L[1];
      box[2][2]=L[2];

      real virial[3][3];
      for(int i=0; i<3; ++i) for(int j=0; j<3; ++j) virial[i][j]=0.0;

      p->cmd("setPositions", pos);
      p->cmd("setForces", f);
      p->cmd("setBox",&box[0][0]);
      p->cmd("setMasses",masses);
      p->cmd("setCharges",charges);
      p->cmd("setVirial", &virial[0][0]);
      p->cmd("getBias",&bias);
      p->cmd("prepareCalc");

      int plumedNeedsEnergy = 0;
      p->cmd("isEnergyNeeded", &plumedNeedsEnergy);
      if (plumedNeedsEnergy) {
        real pot_energy = 0.;
        const interaction::InteractionList& srIL = system.shortRangeInteractions;
        for (size_t j =0; j < srIL.size(); ++j) {
          pot_energy += srIL[j]->computeEnergy();
        }
        pot_energy /= system.comm->size(); // PLUMED defines PE this way.
        p->cmd("setEnergy", &pot_energy);
      }
      p->cmd("performCalc");

      for(auto tp=std::make_pair(0, CellListIterator(realCells));
          tp.first<nreal && !tp.second.isDone();
          ++tp.first, ++tp.second)
        {
          std::copy(&f[tp.first*3], &f[tp.first*3+3], tp.second->force().begin());
        }
    }

    void ExtPlumed::updateStep() {
      step = integrator->getStep()+1;
    }

    /****************************************************
     ** REGISTRATION WITH PYTHON
     ****************************************************/

    void ExtPlumed::registerPython() {

      using namespace espressopp::python;

      class_<ExtPlumed, shared_ptr<ExtPlumed>, bases<Extension> >

        ("integrator_ExtPlumed", init< shared_ptr< System >, python::object, string, string, real, bool>())
        .def("getBias", &ExtPlumed::getBias)
        .def("connect", &ExtPlumed::connect)
        .def("disconnect", &ExtPlumed::disconnect)
        ;
    }
  }
}
